airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -42,35 +61,52 @@ def read_records(
             # On read requests we should always have the catalog available
             raise MissingSchemaError("Expected `json_schema` in the configured catalog but it is missing.")
         parser = self.get_parser(self.config.file_type)
-        for file in self.list_files_for_this_sync(stream_slice):
+        for file in stream_slice["files"]:
+            # only serialize the datetime once
+            file_datetime_string = file.last_modified.strftime("%Y-%m-%dT%H:%M:%SZ")
             try:
                 for record in parser.parse_records(file, self._stream_reader):
                     if not record_passes_validation_policy(self.config.validation_policy, record, schema):
                         logging.warning(f"Record did not pass validation policy: {record}")
                         continue
+                    record[self.ab_last_mod_col] = file_datetime_string
+                    record[self.ab_file_name_col] = file.uri
                     yield stream_data_to_airbyte_message(self.name, record)
+                self._cursor.add_file(file)
             except Exception as exc:
                 raise RecordParseError(f"Error reading records from file: {file.uri}. Is the file valid {self.config.file_type}?") from exc
 
-    @cache
-    def get_json_schema(self) -> Mapping[str, Any]:
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        """
+        Override to return the default cursor field used by this stream e.g: an API entity might always use created_at as the cursor field.
+        :return: The name of the field used as a cursor. If the cursor is nested, return an array consisting of the path to the cursor.
         """
-        Return the JSON Schema for a stream.
+        return self.ab_last_mod_col
 
-        If the user provided a schema, return that. Otherwise, infer it.
+    @cache
+    def get_json_schema(self) -> Mapping[str, Any]:
+        extra_fields = {
+            self.ab_last_mod_col: {"type": "string"},
+            self.ab_file_name_col: {"type": "string"},
+        }
+        schema = self.get_raw_json_schema()
+        schema["properties"] = {**extra_fields, **schema.get("properties", {})}
+        return schema
 
-        Use no more than `_discovery_policy.max_n_files_for_schema_inference` files.
-        """
+    @cache
+    def get_raw_json_schema(self) -> Mapping[str, Any]:
```

Comment:
Is this public? If so, why? The same question applies for `list_files`

Reply:
good catch. I don't think either need to be public (and we should be able to remove `list_files` from the interface)